From: Anders Larsen <al@alarsen.net>
Date: Tue, 26 Jan 2010 18:06:44 +0100
Subject: [PATCH] Fix wxTCPEventHandler to properly check for a valid connection.
 On MSW, wxSOCKET_INPUT events are regularly received even if not data is available.
 Fix wxTCPEventHandler::Client_OnRequest() to check that the connection is not
 being torn down and that there actually are data available to read.
 Fixes #11528.

Signed-off-by: Anders Larsen <al@alarsen.net>
---
 src/common/sckipc.cpp |   16 +++++++++++-----
 1 files changed, 11 insertions(+), 5 deletions(-)

diff --git a/src/common/sckipc.cpp b/src/common/sckipc.cpp
index 0f1173b..9698641 100644
--- a/src/common/sckipc.cpp
+++ b/src/common/sckipc.cpp
@@ -574,12 +574,13 @@ bool wxTCPConnection::Disconnect()
     if ( !GetConnected() )
         return true;
 
-    // Send the disconnect message to the peer.
-    IPCOutput(m_streams).Write8(IPC_DISCONNECT);
-
     if ( m_sock )
     {
         m_sock->Notify(false);
+
+        // Send the disconnect message to the peer.
+        IPCOutput(m_streams).Write8(IPC_DISCONNECT);
+
         m_sock->Close();
     }
 
@@ -707,7 +708,7 @@ void wxTCPEventHandler::HandleDisconnect(wxTCPConnection *connection)
 void wxTCPEventHandler::Client_OnRequest(wxSocketEvent &event)
 {
     wxSocketBase *sock = event.GetSocket();
-    if (!sock)
+    if (!sock || !sock->IsOk())
         return;
 
     wxSocketNotify evt = event.GetSocketEvent();
@@ -715,7 +716,7 @@ void wxTCPEventHandler::Client_OnRequest(wxSocketEvent &event)
         connection = static_cast<wxTCPConnection *>(sock->GetClientData());
 
     // This socket is being deleted; skip this event
-    if (!connection)
+    if (!connection || !connection->GetConnected())
         return;
 
     if ( evt == wxSOCKET_LOST )
@@ -724,6 +725,11 @@ void wxTCPEventHandler::Client_OnRequest(wxSocketEvent &event)
         return;
     }
 
+    // Check if spurious wxSOCKET_INPUT event (happens on MSW)
+    char ch;
+    if (sock->Peek(&ch, 1).LastCount() == 0)
+        return;
+
     // Receive message number.
     wxIPCSocketStreams * const streams = connection->m_streams;
 
-- 
1.5.4.3

